lektor/publisher.py

```diff
@@ -676,10 +691,13 @@ def git(args, **kwargs):
             self.link_artifacts(path)
             self.write_cname(path, target_url)
             for line in git(['add', '-f', '--all', '.']).output:
+                self._check_error(line)
                 yield line
```

Comment:
Maybe something like this:
```python
cmd = git(['add', '-f', '--all', '.'])
lines = []
for line in cmd.output:
    lines.append(line)
    yield line

self._check_error(cmd.returncode, lines)
```

Since we do not know the line that actually displays an error, we should probably store them and pass them to the helper function to check the return code and print all the lines.

Something like that @runfalk ?

Reply:
I think it'd be better to change the implementation of Command to throw exceptions by default if the return code is != 0. That way we don't need the `self._check_error()` calls. We can add a kwarg to the constructor at a later point if we want to suppress errors somewhere else. Currently it's only used for the git command inside this file.

Reply:
I think there is a problem using the return code:
The output is 1 in errors and other lines that are not errors.

@runfalk @goanpeca Any ideas?

<img width="972" alt="Screen Shot 2020-05-26 at 11 52 28 AM" src="https://user-images.githubusercontent.com/14989202/82945612-d5c09180-9f62-11ea-8929-e1a1c4bd3276.png">

Reply:
I assume we want to abort on any failed command? And 0 normally means success when it comes to return codes.

Reply:
I think that what @arendondiosa is saying is that even non failing commands are rteturning a 1 value for the return code.

Reply:
I assumed the 1 after `.....` was the debug printed return code.

Which OS are you testing on @arendondiosa? When I try git on Linux I get a non-zero return code:

```
$ git remote add test http://x
$ git push test master
fatal: kan inte komma åt "http://x/": Could not resolve host: x
ante@Balthazar:~/dev/lektor ((3.1.3))$ echo $?
128
```

Reply:
Testing git commands:
<img width="759" alt="Screen Shot 2020-05-26 at 4 57 09 PM" src="https://user-images.githubusercontent.com/14989202/82954340-22f82f80-9f72-11ea-87af-d5285ea777b0.png">

@runfalk What I mean: I think some commands should return code 0 (e.g. `On branch gh-pages` or` Initial commit`) 🤔 

If we use the validation `return code == 1`, the execution ends with commands that are not errors

Reply:
I think my quick replies may have introduced some confusion. All commands that succeeds will by convention have 0 as its return code. All commands that fail will by convention have a non-zero code, some programs use different codes for different errors while some don't.

The normal way to test would be:

```
is_successful = returncode == 0
```

Reply:
@runfalk This is clear to me

> The normal way to test would be:
> `is_successful = returncode == 0`

What I don't understand is why outputs like `On branch gh-pages` or `Initial commit` have as returncode `1` if these are not errors (I think) 🤔

Reply:
Oh, I see! So *nothing to commit* kind of is an error (at least not the expected case). We can either special case this command or add the `--allow-empty` to allow commits with no changes.

In the image you posted earlier everything fails at the fetch step so nothing after that should succeed I would assume? I assume you intentionally misspelled the repo name to trigger an error? And since the fetch fails I wouldn't expect any command after that to succeed.

Reply:
You're right, in the example above I caused an error to test the outputs.

In this example, the command is successful but there are outputs with return code = 1 (I think these are not errors). So, I think we couldn't use `returncode == 1` to validate errors.

What do you think?

<img width="1440" alt="Screen Shot 2020-05-31 at 3 23 07 PM" src="https://user-images.githubusercontent.com/14989202/83362103-41846f00-a354-11ea-87df-640bd9af3ab6.png">